relay/auth/authenticator.go

```diff
@@ -0,0 +1,159 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	pb "github.com/Layr-Labs/eigenda/api/grpc/relay"
+	"github.com/Layr-Labs/eigenda/core"
+	"sync"
+	"time"
+)
+
+// RequestAuthenticator authenticates requests to the relay service. This object is thread safe.
+type RequestAuthenticator interface {
+	// AuthenticateGetChunksRequest authenticates a GetChunksRequest, returning an error if the request is invalid.
+	// The address is the address of the peer that sent the request. This may be used to cache auth results
+	// in order to save server resources.
+	AuthenticateGetChunksRequest(
+		address string,
+		request *pb.GetChunksRequest,
+		now time.Time) error
+}
+
+// authenticationTimeout is used to track the expiration of an auth.
+type authenticationTimeout struct {
+	clientID   string
+	expiration time.Time
+}
+
+var _ RequestAuthenticator = &requestAuthenticator{}
+
+type requestAuthenticator struct {
+	ics core.IndexedChainState
+
+	// authenticatedClients is a set of client IDs that have been recently authenticated.
+	authenticatedClients map[string]struct{}
+
+	// authenticationTimeouts is a list of authentications that have been performed, along with their expiration times.
+	authenticationTimeouts []*authenticationTimeout
+
+	// authenticationTimeoutDuration is the duration for which an auth is valid.
+	// If this is zero, then auth saving is disabled, and each request will be authenticated independently.
+	authenticationTimeoutDuration time.Duration
+
+	// savedAuthLock is used for thread safe atomic modification of the authenticatedClients map and the
+	// authenticationTimeouts queue.
+	savedAuthLock sync.Mutex
+}
+
+// NewRequestAuthenticator creates a new RequestAuthenticator.
+func NewRequestAuthenticator(
+	ics core.IndexedChainState,
+	authenticationTimeoutDuration time.Duration) RequestAuthenticator {
+
+	return &requestAuthenticator{
+		ics:                           ics,
+		authenticatedClients:          make(map[string]struct{}),
+		authenticationTimeouts:        make([]*authenticationTimeout, 0),
+		authenticationTimeoutDuration: authenticationTimeoutDuration,
+	}
+}
+
+func (a *requestAuthenticator) AuthenticateGetChunksRequest(
+	address string,
+	request *pb.GetChunksRequest,
+	now time.Time) error {
+
+	if a == nil {
+		// do not enforce auth if the authenticator is nil
+		return nil
+	}
+
+	if a.isAuthenticationStillValid(now, address) {
+		// We've recently authenticated this client. Do not authenticate again for a while.
+		return nil
+	}
+
+	blockNumber, err := a.ics.GetCurrentBlockNumber()
+	if err != nil {
+		return fmt.Errorf("failed to get current block number: %w", err)
+	}
+	operators, err := a.ics.GetIndexedOperators(context.Background(), blockNumber)
```

Comment:
There are 2 RPC calls here, it'd be helpful to cache the operator state

Reply:
caching added